saliency/tf1/base.py

```diff
@@ -0,0 +1,194 @@
+      feed_dict: (Optional) feed dictionary to pass to the session.run call.
+    """
+    raise NotImplementedError('A derived class should implemented GetMask()')
+
+  def GetSmoothedMask(self,
+                      x_value,
+                      feed_dict={},
+                      stdev_spread=.15,
+                      nsamples=25,
+                      magnitude=True,
+                      **kwargs):
+    """Returns a mask that is smoothed with the SmoothGrad method.
+
+    Args:
+      x_value: Input value, not batched.
+      feed_dict: (Optional) feed dictionary to pass to the session.run call.
+      stdev_spread: Amount of noise to add to the input, as fraction of the
+        total spread (x_max - x_min). Defaults to 15%.
+      nsamples: Number of samples to average across to get the smooth gradient.
+      magnitude: If true, computes the sum of squares of gradients instead of
+        just the sum. Defaults to true.
+      **kwargs: Additional keyword arguments to be passed to GetMask.
+    """
+    stdev = stdev_spread * (np.max(x_value) - np.min(x_value))
+
+    total_gradients = np.zeros_like(x_value)
+    for _ in range(nsamples):
+      noise = np.random.normal(0, stdev, x_value.shape)
+      x_plus_noise = x_value + noise
+      grad = self.GetMask(x_plus_noise, feed_dict, **kwargs)
+      if magnitude:
+        total_gradients += (grad * grad)
+      else:
+        total_gradients += grad
+
+    return total_gradients / nsamples
+
+
+class TF1CoreSaliency(TF1Saliency):
+  """Base class for TF1Saliency methods that use CoreSaliency methods.
+
+  Alone, this class doesn't do anything.
+  """
+
+  def __init__(self, graph, session, y=None, x=None, conv_layer=None):
+    """TF1-specific class to create saliency masks using CoreSaliency.
+
+    Args:
+      graph: The TensorFlow graph to evaluate masks on.
+      session: The current TensorFlow session.
+      y: The output tensor to compute the SaliencyMask against. This tensor
+          should be of size 1.
+      x: The input tensor to compute the SaliencyMask against. The outer
+          dimension should be the batch size.
+      conv_layer: The convolution layer tensor of the model. The outer
+          dimension should be the batch size.
+    """
+
+    super(TF1CoreSaliency, self).__init__(graph, session, y, x)
+    self.call_model_function = utils.create_tf1_call_model_function(
+        graph, session, y, x, conv_layer)
+
+  def validate_xy_tensor_shape(self, x_steps=1, batch_size=1):
+    """Validates the shapes of x and y tensors with respect to batch_size.
+
+    Args:
+      x_steps: Number of steps computed as part of the full method.
+      batch_size: Maximum batch_size to split up the x_steps.
+
+    Raises:
+        ValueError: If x tensor shape is incompatible with batching parameters.
+                    If y tensor shape is incompatible with batching parameters.
+    """
+    target_size = 1
+    batch_remainder = x_steps % batch_size
```

Comment:
I think that it is possible for these to be different. If there is no remainder, the input tensor can have an outermost dimension equal to that batch size (e.g. [32,299,299,3] instead of [None,299,299,3]), but if there's a remainder it won't work with a fixed input tensor shape. It's a niche case, so we could remove this validation and always ask for the size to be None.